runtime/caml/signals.h

```diff
@@ -45,6 +45,27 @@ CAMLextern value caml_process_pending_actions_exn (void);
 
 #ifdef CAML_INTERNALS
 CAMLextern intnat volatile caml_pending_signals[];
+
+/* When an action is pending, either [caml_something_to_do] is 1, or
+   there is a function currently running which will end by either
+   executing all actions, or set [caml_something_to_do] back to 1. It
+   is 0 otherwise.
+
+   FIXME:
+
+    * We could get into caml_process_pending_actions when
+      caml_something_to_do is seen as set but not
+      caml_pending_signals, making us miss the signal.
+
+    * If there are two different callbacks (say, a signal and a
+      finaliser) arriving at the same time, then we set
+      caml_something_to_do to 0 when starting the first one while the
+      second one is still waiting. We may want to run the second one
+      if the first one is taking long.
```

Comment:
Not sure this deserves a FIXME. The current solution is rather good, and I don't think there is anything better we could do.

Reply:
IIUC, in these situations, instead of reacting at the next allocation, it react on the order of 10^5 to 10^6 allocations later. For finalisers the total latency remains on the same order, but for signals this is bad. In particular, have a look at the part of https://github.com/ocaml-multicore/ocaml-multicore/issues/187#issuecomment-532521642 which says "guaranteed interrupt at fixed delays... Based on the maximum latency of interrupt required...". From their point of view of guaranteed latency, this is still buggy.

The second bullet point can be solved if we call signal handlers first, what do you think about it?

It would be nice to fix the first one as well.

Reply:
> The second bullet point can be solved if we call signal handlers first, what do you think about it?

Alright, let's do that. Note, however, that the issue is still there for finalisers or memprof. That's sort of an unsolvable problem.

Anyway, for both issues, they will only occur at very rare occasions, and we have a workaround. And it's unlikely we will find a "fix" any time. Hence I don't think the "FIXME" is deserved.

But that's a detail, I don't care, let's move forward.

Reply:
The second bullet point has been fixed. Signal handlers are now safe from this (newly-introduced, AFAIU) race. It is no longer a FIXME, because it is not a bug for finalisers (their latency remains on the same order of magnitude), and it is not a bug for memprof callbacks either, because although we want them to run ASAP, they are served on a best-effort basis.

I keep the first item as a FIXME because it is a bug if we take into account code that tries to allocate little, or if there are latency guarantees as assumed in the work on better safe points. If it is a deviation to the intended behaviour and it does not appear easily solvable with the current mechanism, then I think a FIXME is the best option.